### PR TITLE
[Core] Unrevert "Add retry exception allowlist for user-defined filtering of retryable application-level errors."

### PR DESCRIPTION
**Pull request by clarkzinzow**
This reverts commit cf7305a2c911f98c35edd4a5c5c0ba1fa4de0cd2, and unreverts https://github.com/ray-project/ray/pull/25896.

This was reverted due to a failing Windows test: https://github.com/ray-project/ray/issues/26287

We can merge once the failing Windows test (and all other relevant tests) pass.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -1559,6 +1635,7 @@ cdef class CoreWorker:
                     max_retries, retry_exceptions,
                     c_scheduling_strategy,
                     debugger_breakpoint,
+                    serialized_retry_exception_allowlist,
```
Comment (mattip) [inline python/ray/_raylet.pyx:1638]: I think this change is what is causing windows compilation to fail with 
```
Error compiling Cython file:
------------------------------------------------------------
...
                return_refs = CCoreWorkerProcess.GetCoreWorker().SubmitTask(
                    ray_function, args_vector, task_options,
                    max_retries, retry_exceptions,
                    c_scheduling_strategy,
                    debugger_breakpoint,
                    serialized_retry_exception_allowlist,
                   ^
------------------------------------------------------------

python\ray\_raylet.pyx:1638:20: Coercion from Python not allowed without the GIL

```

**Comment by mattip** [comment]
There is still a [compilation error in windows](https://buildkite.com/ray-project/ray-builders-pr/builds/40107#01823baa-cf4d-44f9-95f6-57d442e6c39e/2227-2972) where a C++ override is not found.

**Comment by clarkzinzow** [comment]
@mattip Yep I know, just pushed up another fix.

**Comment by clarkzinzow** [comment]
I'll ping the Slack thread once this is ready to be looked at.

**Comment by mattip** [comment]
There is one failing test on windows: [test_output::test_fail_importing_actor](https://buildkite.com/ray-project/ray-builders-pr/builds/40145#01823bfb-f5d2-4bdf-8fc6-2b0f1ca8e6f9/3417-7388). It hangs, where the test is designed to exit with an error. I can reproduce this single test hang locally via `pytest python\ray\tests\test_output.py -k test_fail_importing_actor` The problem seems to be that the error does not close the ray process and allow the test to continue, rather it hangs. 

The following test `test_fail_importing_task`, which is very similar, does not hang and passes when run separately via pytest. The import error is printed to stdout, where in `test_fail_importing_actor` the error only shows up in the `rayley.out` log.

**Comment by mattip** [comment]
Windows [now fails](https://buildkite.com/ray-project/ray-builders-pr/builds/40807#01824c75-b922-452d-b9c8-bfe3e773ebe0) `test_object_spilling` where it passed before.

**Comment by mattip** [comment]
Windows now passed. Some other runs newly failed

**Comment by mattip** [comment]
I have started to look at the [test_output::test_fail_importing_actor](https://buildkite.com/ray-project/ray-builders-pr/builds/40145#01823bfb-f5d2-4bdf-8fc6-2b0f1ca8e6f9/3417-7388) failure. Here are a few observations:
- the test runs a string via `run_string_as_driver_nonblocking`. If instead I create a file with the string and run it via `python testme.py` the process does not hang.
- In the log files, the actor crash with the desired `ModuleNotFoundError` reaches the `gcs_server.out` log, and the `gcs_server` reports that it is stopped.
- In the `raylet.out` log, the raylet process shuts down.
- I still see a `raylet.exe` process in TaskManager, even minutes after the shutdown.
- There is an obscure error being raised from `ActorHandle.__del__`. `ray._private` is None, as if the interpreter is being shut down. This means the `ray._private.worker.global_worker.core_worker.remove_actor_handle_reference()` method will never be called. Perhaps that is causing a deadlock somewhere and preventing the interpreter from completely closing?
- `ray stack` does not work on windows.

**Comment by clarkzinzow** [comment]
Thank you for looking into that test failure!

It looks like all test failures are unrelated, I'm going to merge latest master and kick off a new build. If the test failures are determined to be unrelated, this should be good to merge, cc @ericl and @jjyao as the reviewers of the original PR.

**Review by ericl — APPROVED** [review]

**Comment by clarkzinzow** [comment]
@mattip @ericl Bumping `test_iter.py` from the medium tests to the large test, since it was already getting close to the medium-test timeout on Windows before this PR, and this PR appeared to push it over the line.

**Comment by clarkzinzow** [comment]
Windows tests are all passing now, will merge once MacOS build fails.

**Comment by clarkzinzow** [comment]
All MacOS builds passed on the previous run, and the only change in the last commit was bumping the timeout for a test (no code changes), so going to merge under the assumption that the MacOS build won't break because of that test timeout bump.

**Pull request merged**